src/main/java/com/google/sps/data/DatastoreClient.java

```diff
@@ -0,0 +1,122 @@
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import com.google.sps.data.Attendee;
+import com.google.sps.data.AttendeeInterface;
+import com.google.sps.data.Session;
+import com.google.sps.data.SessionInterface;
+import com.google.sps.data.Instance;
+import com.google.sps.data.InstanceInterface;
+
+/** Class that reads from and writes to Datastore. */
+public class DatastoreClient implements DatastoreClientInterface {
+  private final DatastoreService datastoreService = 
+    DatastoreServiceFactory.getDatastoreService();
+  private final String INSTANCE_STATE_TERMINATED = "TERMINATED";
+
+  public void insertOrUpdateAttendee(AttendeeInterface attendee) {
+    datastoreService.put(attendee.toEntity());
+  }
+
+  public void insertOrUpdateInstance(InstanceInterface instance) {
+    datastoreService.put(instance.toEntity());
+  }
+
+  public void insertOrUpdateSession(SessionInterface session) {
+    datastoreService.put(session.toEntity());
+  }
+
+  public Optional<SessionInterface> getSession(String sessionId) {
+    Query getSession = new Query(EntityConstants.SessionEntity.TABLE_NAME)
+      .setFilter(new FilterPredicate(EntityConstants.SessionEntity.SESSION_ID,
+      FilterOperator.EQUAL, sessionId));
+    PreparedQuery pq = datastoreService.prepare(getSession);
+    Entity sessionEntity = pq.asSingleEntity();
+    return Optional.of(Session.fromEntity(sessionEntity));
+  }
+
+  public Optional<InstanceInterface> getInstance(String instanceName) {
+    Query getInstance = new Query(EntityConstants.InstanceEntity.TABLE_NAME)
+      .setFilter(new FilterPredicate
+      (EntityConstants.InstanceEntity.INSTANCE_NAME, 
+      FilterOperator.EQUAL, instanceName));
+    PreparedQuery pq = datastoreService.prepare(getInstance);
+    Entity instanceEntity = pq.asSingleEntity();
+    return Optional.of(Instance.fromEntity(instanceEntity));
+   }
+
+  public Optional<AttendeeInterface> getAttendee(String screenName) {
+    Query getAttendee = new Query(EntityConstants.AttendeeEntity.TABLE_NAME)
+      .setFilter(new FilterPredicate
+      (EntityConstants.AttendeeEntity.SCREEN_NAME,
+      FilterOperator.EQUAL, screenName));
+    PreparedQuery pq = datastoreService.prepare(getAttendee);
+    Entity attendeeEntity = pq.asSingleEntity();
+    return Optional.of(Attendee.fromEntity(attendeeEntity));
+   }
+
+  public void deleteAttendee(String screenName) {
```

Comment:
how can you delete an attendee based off just a name? names are unique for a session but not for everything. same for get attendee too

Reply:
yeah good catch. You'll need the session information. Which also means that you cannot create an Attendee with the key being their name because their name isnt globally unique. The key must be a combination of their name and the session they are in